keras_cv/models/segmentation/fcn.py

```diff
@@ -0,0 +1,268 @@
+# Copyright 2022 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+from tensorflow import keras
+
+BACKBONE_CONFIG = {
+    "vgg16": {"BLOCK3": 11, "BLOCK4": 15, "BLOCK5": 19, "DENSE_START": -3},
+    "vgg19": {"BLOCK3": 12, "BLOCK4": 17, "BLOCK5": 22, "DENSE_START": -3},
+}
+
+
+class FCN(tf.keras.models.Model):
+    """A segmentation model based on the Fully Convolutional Network introduced by Long et. al.
+
+    Args:
+        classes: int, the number of classes for the segmentation model. Note that
+            the classes doesn't contain the background class, and the classes
+            from the data should be represented by integers with range
+            [0, classes).
+        backbone: a backbone network for the model. Can be a `tf.keras.models.Model`
+            instance. The supported pre-defined backbone models are:
+            1. "vgg16", a VGG16 model
+            2. "vgg19", a VGG19 model
+            Defaults to 'vgg16'.
+            Note: If a custom `tf.keras.models.Model` is passed, then only the
+            `Conv2D`, `MaxPooling2D` and `Dense` layers are extracted from it
+            to make the custom backbone.
+        model_architecture: str, defines the model architecture based on the
+            implementation details present in the paper. The supported
+            architectures are:
+            1. 'fcn8s', a FCN-8S definition
+            2. 'fcn16s', a FCN-16S definition
+            3. 'fcn32s', a FCN-32S definition
+            Defaults to 'fcn8s'.
+
+    """
+
+    def _init_(self, classes, backbone="vgg16", model_architecture=None, **kwargs):
+        super(FCN, self)._init_(**kwargs)
+
+        if isinstance(backbone, str):
+
+            if backbone == "vgg16":
```

Comment:
Should the backbones have weights loaded in?
If not - we should probably prefer using `keras_cv.models` (VGG19 exists, VGG16 is in a PR) instead, since `keras.applications` will eventually be deprecated.

Reply:
That's actually fine either way, weights loaded in or not. If `keras.applications` is headed towards deprecation, then sure, I'll refactor to use `keras_cv.models` instead. VGG16 has also been merged as far as I know, as per the discussions in #1164 so shouldn't be too big a refactor

Reply:
`keras.applications` will eventually be deprecated, when these models get mature enough. If we already can use KCV without any blockage (such as weights being required, but not existing yet), IMO, we should.

KCV VGG16 doesn't have weights currently, so training FCN (if it requires backbone weights) might be more difficult, though.

Reply:
As such, there is no code that requires the presence of weights. This will only end up training the network fully with random weights for now, instead of using pre-initialized weights. In the case of VGG19, it maye end up being initialized with the weights and hence, possibly quicker.

Reply:
For the time being, we can utilise kcv vgg16 without weights; once the weights have been uploaded, we can load the backbone weights.

Reply:
Have made the change. Resolving for now, can reopen if needed.